### ISSUE TITLE
SPARQL variable assignment fails instead of producing unbound variable

### ISSUE THREAD

**Issue by jindrichmynarz** [RDFLib/rdflib#1873]
Using the SPARQL query engine of rdflib 6.1.1, when an assignment's operand is an error (e.g., running a function on an unbound variable), an error is thrown, for example:

```py
from rdflib import Graph

[r for r in Graph().query('SELECT (concat("", sha1(?x)) AS ?y) WHERE {}')]

# AttributeError: 'NotBoundError' object has no attribute 'datatype'
```

However, the [SPARQL 1.1 specification](https://www.w3.org/TR/sparql11-query/#assignment) expects that the assigned variable would remain unbound in such case:

> If the evaluation of the expression produces an error, the variable remains unbound for that solution but the query evaluation continues.

This is also what Apache Jena 4.4.0 does.

**Comment by mielvds** [comment]
THis might be related: 

```
BIND ('https://data.hetarchief.be/id/organisatie/' AS ?ns)
```
results into `AttributeError: 'SPARQLError' object has no attribute 'datatype'`

```
BIND (URI('https://data.hetarchief.be/id/organisatie/') AS ?ns)
``` 
results into `AttributeError: 'URIRef' object has no attribute 'datatype'`

**Comment by gjhiggins** [comment]
Hmm, perhaps this just needs a guard.

**Comment by aucampia** [comment]
- https://github.com/RDFLib/rdflib/pull/2013 should have fixed this and it was included in RDFLib 6.2.0

Closing this issue but if it still persists please re-open it.

**Comment by jindrichmynarz** [comment]
Thanks! I can no longer reproduce the above issue when using rdflib version 6.2.0.
